### ISSUE TITLE
Educational purpose: Implementation question

### ISSUE THREAD

**Issue by sfc-gh-japatel** [pinterest/secor#1641]
Hello there, 
This question is just for implementation purpose. 
In readme file, it is written that 
`as long as Kafka is not dropping messages (e.g., due to aggressive cleanup policy) before Secor is able to read them, it is guaranteed that each message will be saved in exactly one S3 file. This property is not compromised by the notorious temporal inconsistency of S3 caused by the eventual consistency model,`

Although, doesn't it also depend on what is the underlining implementation for uploadManager? 
Does `Hadoops3uploadmanager` provide strong consistency? 

Thanks

**Comment by HenryCaiHaiying** [comment]
Exact-once upload has not much to do with which upload manager, it depends
on the following things:
1. The file name generation is deterministic: kafka-partition-number +
begin-kafka-offset
2. S3 upload first then commit kafka consumer offset
3. In the case when S3 upload succeeds but kafka consumer offset commit
fails, the next secor worker will continue working on this partition and
re-upload the whole thing again (starting with the same begin_offset since
that offset was not committed to kafka), the file would still be named the
same and it will overwrite the the existing file on S3

**Comment by sfc-gh-japatel** [comment]
Thanks Henry, does any of this operation do a get before put? Looks like if you are just replacing the file in 3rd step then there might not be a need for get.

**Comment by HenryCaiHaiying** [comment]
There is no get, it's a file replacing operation on S3.
